Seg fault applying Node using KDD if BGP section is empty.

<!--- Provide a general summary of the issue in the Title above -->

If you apply Node resource configuration with an empty BGP section using KDD backend, we hit a seg fault in libcalico-go.

```
dist/calicoctl-darwin-amd64 apply -f temp-node.yaml 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1dbba6b]

goroutine 1 [running]:
github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources.mergeCalicoK8sNode(0xc4207b8840, 0xc4200bb600, 0x10, 0x0, 0x0)
	/go/src/github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources/node_conversion.go:70 +0x3b
github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources.(*nodeClient).Update(0xc42017e630, 0xc4207b8880, 0xc4208432b0, 0x10661b2, 0xc4202d3fc0)
	/go/src/github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources/node.go:56 +0x1a5
github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources.(*nodeClient).Apply(0xc42017e630, 0xc4207b8880, 0x38, 0x2039c20, 0xc420474000)
	/go/src/github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources/node.go:84 +0x39
github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources.(*retryWrapper).Apply(0xc4202ce930, 0xc4207b8880, 0xc420843388, 0x1, 0x1)
	/go/src/github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources/retrywrapper.go:85 +0x64
github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.(*KubeClient).Apply(0xc4201fe2c0, 0xc4207b8880, 0xc420012b90, 0xc4207b8880, 0x0)
	/go/src/github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/k8s.go:389 +0x69b
github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/client.(*Client).apply(0xc4202f8ee0, 0x29b51c0, 0xc420012b90, 0x29c2ec0, 0xc42017e648, 0x1fcc380, 0x1)
	/go/src/github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/client/client.go:266 +0xb0
github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/client.(*nodes).Apply(0xc42017e648, 0xc4204ee7d0, 0x26c94e27526c0c26, 0x2a26e20, 0x2e1e000)
	/go/src/github.com/projectcalico/calicoctl/vendor/github.com/projectcalico/libcalico-go/lib/client/node.go:74 +0xbf
github.com/projectcalico/calicoctl/calicoctl/resourcemgr.init.4.func1(0xc4202f8ee0, 0x29b51c0, 0xc4201633b0, 0xc42033a1e0, 0x20e6340, 0xc4207716d8, 0x1e48bca)
	/go/src/github.com/projectcalico/calicoctl/calicoctl/resourcemgr/node.go:37 +0x132
github.com/projectcalico/calicoctl/calicoctl/resourcemgr.resourceHelper.Apply(0x211ca75, 0x4, 0x211bbb2, 0x2, 0x29cc180, 0x2030c20, 0xc4203141d0, 0x1, 0x1, 0xc420212980, ...)
	/go/src/github.com/projectcalico/calicoctl/calicoctl/resourcemgr/resourcemgr.go:320 +0x50
github.com/projectcalico/calicoctl/calicoctl/resourcemgr.(*resourceHelper).Apply(0xc42033a1e0, 0xc4202f8ee0, 0x29b51c0, 0xc4201633b0, 0xc420771748, 0x10ff7df, 0xc420771ae8, 0x0)
	<autogenerated>:4 +0xb1
github.com/projectcalico/calicoctl/calicoctl/commands.executeResourceAction(0xc420360c30, 0xc4202f8ee0, 0x29b51c0, 0xc4201633b0, 0x0, 0x0, 0x0, 0x1, 0xc42017e4e0)
	/go/src/github.com/projectcalico/calicoctl/calicoctl/commands/resources.go:287 +0x345
github.com/projectcalico/calicoctl/calicoctl/commands.executeConfigCommand(0xc420360c30, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/projectcalico/calicoctl/calicoctl/commands/resources.go:266 +0x6d4
github.com/projectcalico/calicoctl/calicoctl/commands.Apply(0xc4203fc6c0, 0x3, 0x3)
	/go/src/github.com/projectcalico/calicoctl/calicoctl/commands/apply.go:87 +0xf7
main.main()
	/go/src/github.com/projectcalico/calicoctl/calicoctl/calicoctl.go:80 +0x6d5
```

## Expected Behavior
BGP configuration should be removed from the node.

## Current Behavior
libcalico-go seg faults when applying configuration with an empty BGP section.

## Steps to Reproduce (for bugs)
Spotted using calicoctl.

```
dist/calicoctl-darwin-amd64 apply -f temp-node.yaml 
cat temp-node.yaml 
- apiVersion: v1
  kind: node
  metadata:
    name: rlb-test-kdd-001
  spec:
    bgp:
```

--- a/lib/backend/k8s/conversion_test.go
+++ b/lib/backend/k8s/conversion_test.go
@@ -25,9 +25,6 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 	k8sapi "k8s.io/client-go/pkg/api/v1"
 	extensions "k8s.io/client-go/pkg/apis/extensions/v1beta1"
-
-	"github.com/projectcalico/libcalico-go/lib/backend/k8s/resources"
-	"github.com/projectcalico/libcalico-go/lib/net"
 )
 
 var _ = Describe("Test parsing strings", func() {
@@ -686,74 +683,5 @@ var _ = Describe("Test Namespace conversion", func() {
 			Expect(inboundRules[0]).To(Equal(model.Rule{Action: "allow"}))
 			Expect(outboundRules[0]).To(Equal(model.Rule{Action: "allow"}))
 		})
-	})
-})
-
-var _ = Describe("Test Node conversion", func() {
-
-	It("should parse a k8s Node to a Calico Node", func() {
-		l := map[string]string{"net.beta.kubernetes.io/role": "master"}
-		node := k8sapi.Node{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:            "TestNode",
-				Labels:          l,
-				ResourceVersion: "1234",
-				Annotations: map[string]string{
-					"projectcalico.org/IPv4Address": "172.17.17.10/24",
-					"projectcalico.org/ASNumber":    "2546",
-				},
-			},
-			Status: k8sapi.NodeStatus{
-				Addresses: []k8sapi.NodeAddress{
-					k8sapi.NodeAddress{
-						Type:    k8sapi.NodeInternalIP,
-						Address: "172.17.17.10",
-					},
-					k8sapi.NodeAddress{
-						Type:    k8sapi.NodeExternalIP,
-						Address: "192.168.1.100",
-					},
-					k8sapi.NodeAddress{
-						Type:    k8sapi.NodeHostName,
-						Address: "172-17-17-10",
-					},
-				},
-			},
-			Spec: k8sapi.NodeSpec{},
-		}
-
-		n, err := resources.K8sNodeToCalico(&node)
-		Expect(err).NotTo(HaveOccurred())
-
-		// Ensure we got the correct values.
-		felixAddress := *n.Value.(*model.Node).FelixIPv4
-		bgpAddress := *n.Value.(*model.Node).BGPIPv4Addr
-		bgpNet := *n.Value.(*model.Node).BGPIPv4Net
-		labels := n.Value.(*model.Node).Labels
-		asn := n.Value.(*model.Node).BGPASNumber
-
-		ip, ipNet, _ := net.ParseCIDR("172.17.17.10/24")
-
-		Expect(felixAddress).To(Equal(*ip))
-		Expect(bgpAddress).To(Equal(*ip))
-		Expect(bgpNet).To(Equal(*ipNet))
-		Expect(labels).To(Equal(l))
-		Expect(asn.String()).To(Equal("2546"))
-	})
-
-	It("should error on an invalid IP", func() {
-		l := map[string]string{"net.beta.kubernetes.io/role": "master"}
-		node := k8sapi.Node{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:            "TestNode",
-				Labels:          l,
-				ResourceVersion: "1234",
-				Annotations:     map[string]string{"projectcalico.org/IPv4Address": "172.984.12.5/24"},
-			},
-			Spec: k8sapi.NodeSpec{},
-		}
-
-		_, err := resources.K8sNodeToCalico(&node)
-		Expect(err).To(HaveOccurred())
 	})
 })

--- a/lib/backend/k8s/resources/node.go
+++ b/lib/backend/k8s/resources/node.go
@@ -25,6 +25,11 @@ import (
 	"k8s.io/client-go/rest"
 )
 
+const (
+	nodeBgpIpv4CidrAnnotation = "projectcalico.org/IPv4Address"
+	nodeBgpAsnAnnotation = "projectcalico.org/ASNumber"
+)
+
 func NewNodeClient(c *kubernetes.Clientset, r *rest.RESTClient) K8sResourceClient {
 	return &retryWrapper{
 		client: &nodeClient{

--- a/lib/backend/k8s/resources/node_conversion.go
+++ b/lib/backend/k8s/resources/node_conversion.go
@@ -67,9 +67,11 @@ func K8sNodeToCalico(node *kapiv1.Node) (*model.KVPair, error) {
 func mergeCalicoK8sNode(calicoNode *model.Node, k8sNode *kapiv1.Node) (*kapiv1.Node, error) {
 	// In order to make sure we always end up with a CIDR that has the IP and not just network
 	// we assemble the CIDR from BGPIPv4Addr and BGPIPv4Net.
-	subnet, _ := calicoNode.BGPIPv4Net.Mask.Size()
-	ipCidr := fmt.Sprintf("%s/%d", calicoNode.BGPIPv4Addr.String(), subnet)
-	k8sNode.Annotations["projectcalico.org/IPv4Address"] = ipCidr
+	if calicoNode.BGPIPv4Net != nil {
+		subnet, _ := calicoNode.BGPIPv4Net.Mask.Size()
+		ipCidr := fmt.Sprintf("%s/%d", calicoNode.BGPIPv4Addr.String(), subnet)
+		k8sNode.Annotations["projectcalico.org/IPv4Address"] = ipCidr
+	}
 
 	// Don't set the ASNumber if it is nil, and ensure it does not exist in k8s.
 	if calicoNode.BGPASNumber != nil {

--- a/lib/backend/k8s/resources/node_conversion.go
+++ b/lib/backend/k8s/resources/node_conversion.go
@@ -35,7 +35,7 @@ func K8sNodeToCalico(node *kapiv1.Node) (*model.KVPair, error) {
 	calicoNode := model.Node{}
 	calicoNode.Labels = node.Labels
 	annotations := node.ObjectMeta.Annotations
-	cidrString, ok := annotations["projectcalico.org/IPv4Address"]
+	cidrString, ok := annotations[nodeBgpIpv4CidrAnnotation]
 	if ok {
 		ip, cidr, err := net.ParseCIDR(cidrString)
 		if err != nil {
@@ -47,7 +47,7 @@ func K8sNodeToCalico(node *kapiv1.Node) (*model.KVPair, error) {
 		calicoNode.BGPIPv4Net = cidr
 	}
 
-	asnString, ok := annotations["projectcalico.org/ASNumber"]
+	asnString, ok := annotations[nodeBgpAsnAnnotation]
 	if ok {
 		asn, err := numorstring.ASNumberFromString(asnString)
 		if err != nil {
@@ -70,14 +70,16 @@ func mergeCalicoK8sNode(calicoNode *model.Node, k8sNode *kapiv1.Node) (*kapiv1.N
 	if calicoNode.BGPIPv4Net != nil {
 		subnet, _ := calicoNode.BGPIPv4Net.Mask.Size()
 		ipCidr := fmt.Sprintf("%s/%d", calicoNode.BGPIPv4Addr.String(), subnet)
-		k8sNode.Annotations["projectcalico.org/IPv4Address"] = ipCidr
+		k8sNode.Annotations[nodeBgpIpv4CidrAnnotation] = ipCidr
+	} else {
+		delete(k8sNode.Annotations, nodeBgpIpv4CidrAnnotation)
 	}
 
 	// Don't set the ASNumber if it is nil, and ensure it does not exist in k8s.
 	if calicoNode.BGPASNumber != nil {
-		k8sNode.Annotations["projectcalico.org/ASNumber"] = calicoNode.BGPASNumber.String()
+		k8sNode.Annotations[nodeBgpAsnAnnotation] = calicoNode.BGPASNumber.String()
 	} else {
-		delete(k8sNode.Annotations, "projectcalico.org/ASNumber")
+		delete(k8sNode.Annotations, nodeBgpAsnAnnotation)
 	}
 
 	return k8sNode, nil

--- a/lib/backend/k8s/resources/node_conversion_test.go
+++ b/lib/backend/k8s/resources/node_conversion_test.go
@@ -1,0 +1,135 @@
+package resources
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sapi "k8s.io/client-go/pkg/api/v1"
+
+	"github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/numorstring"
+)
+
+var _ = Describe("Test Node conversion", func() {
+
+	It("should parse a k8s Node to a Calico Node", func() {
+		l := map[string]string{"net.beta.kubernetes.io/role": "master"}
+		node := k8sapi.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            "TestNode",
+				Labels:          l,
+				ResourceVersion: "1234",
+				Annotations: map[string]string{
+					nodeBgpIpv4CidrAnnotation: "172.17.17.10/24",
+					nodeBgpAsnAnnotation:    "2546",
+				},
+			},
+			Status: k8sapi.NodeStatus{
+				Addresses: []k8sapi.NodeAddress{
+					k8sapi.NodeAddress{
+						Type:    k8sapi.NodeInternalIP,
+						Address: "172.17.17.10",
+					},
+					k8sapi.NodeAddress{
+						Type:    k8sapi.NodeExternalIP,
+						Address: "192.168.1.100",
+					},
+					k8sapi.NodeAddress{
+						Type:    k8sapi.NodeHostName,
+						Address: "172-17-17-10",
+					},
+				},
+			},
+			Spec: k8sapi.NodeSpec{},
+		}
+
+		n, err := K8sNodeToCalico(&node)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Ensure we got the correct values.
+		felixAddress := *n.Value.(*model.Node).FelixIPv4
+		bgpAddress := *n.Value.(*model.Node).BGPIPv4Addr
+		bgpNet := *n.Value.(*model.Node).BGPIPv4Net
+		labels := n.Value.(*model.Node).Labels
+		asn := n.Value.(*model.Node).BGPASNumber
+
+		ip, ipNet, _ := net.ParseCIDR("172.17.17.10/24")
+
+		Expect(felixAddress).To(Equal(*ip))
+		Expect(bgpAddress).To(Equal(*ip))
+		Expect(bgpNet).To(Equal(*ipNet))
+		Expect(labels).To(Equal(l))
+		Expect(asn.String()).To(Equal("2546"))
+	})
+
+	It("should error on an invalid IP", func() {
+		l := map[string]string{"net.beta.kubernetes.io/role": "master"}
+		node := k8sapi.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            "TestNode",
+				Labels:          l,
+				ResourceVersion: "1234",
+				Annotations:     map[string]string{nodeBgpIpv4CidrAnnotation: "172.984.12.5/24"},
+			},
+			Spec: k8sapi.NodeSpec{},
+		}
+
+		_, err := K8sNodeToCalico(&node)
+		Expect(err).To(HaveOccurred())
+	})
+
+	It("Should parse and remove BGP info when given Calico Node with empty BGP spec", func() {
+		l := map[string]string{"net.beta.kubernetes.io/role": "master"}
+		k8sNode := &k8sapi.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            "TestNode",
+				Labels:          l,
+				ResourceVersion: "1234",
+				Annotations: map[string]string{
+					nodeBgpIpv4CidrAnnotation: "172.17.17.10/24",
+					nodeBgpAsnAnnotation:    "2546",
+				},
+			},
+			Spec: k8sapi.NodeSpec{},
+		}
+
+		calicoNode := &model.Node{}
+
+		newK8sNode, err := mergeCalicoK8sNode(calicoNode, k8sNode)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(newK8sNode.Annotations).NotTo(HaveKey(nodeBgpIpv4CidrAnnotation))
+		Expect(newK8sNode.Annotations).NotTo(HaveKey(nodeBgpAsnAnnotation))
+	})
+
+	It("Should merger Calico Nodes into K8s Nodes", func() {
+		l := map[string]string{"net.beta.kubernetes.io/role": "master"}
+		k8sNode := &k8sapi.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            "TestNode",
+				Labels:          l,
+				ResourceVersion: "1234",
+				Annotations: make(map[string]string),
+			},
+			Spec: k8sapi.NodeSpec{},
+		}
+
+		ip, cidr, _ := net.ParseCIDR("172.17.17.10/24")
+		asn, _ := numorstring.ASNumberFromString("2456")
+
+		calicoNode := &model.Node{
+			BGPIPv4Net: cidr,
+			FelixIPv4: ip,
+			BGPIPv4Addr: ip,
+			BGPASNumber: &asn,
+		}
+
+		newK8sNode, err := mergeCalicoK8sNode(calicoNode, k8sNode)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(newK8sNode.Annotations).To(HaveKeyWithValue(nodeBgpIpv4CidrAnnotation, "172.17.17.10/24"))
+		Expect(newK8sNode.Annotations).To(HaveKeyWithValue(nodeBgpAsnAnnotation, "2456"))
+	})
+})
+

